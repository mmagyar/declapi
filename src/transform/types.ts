import Ajv from 'ajv'
import { ObjectType } from 'yaschva'
import { Elastic, KeyValue } from '../DataDriverTypes.js'
import { HttpMethods, SearchTypes } from '../globalTypes.js'
import { dirname } from 'path'
import { fileURLToPath } from 'url'

export type AuthType = (string | {createdBy: boolean})[] | boolean
export type ManageableFields ={ createdBy?: boolean }

export type Contract = {
  name: string;
  type?: HttpMethods;
  authentication: AuthType;
  manageFields: ManageableFields;
  arguments: ObjectType;
  returns: ObjectType;
  description?: string;
};
export type CrudAuthAll = {
  get: AuthType;
  put: AuthType;
  post: string[] | boolean;
  delete: AuthType;
};

export type CrudAuthSome = {
  get: AuthType;
  modify: AuthType;
  delete?: AuthType;
};

export type CrudContract = {
  name: string;
  methods?: { get?:boolean, post?:boolean, put?: boolean, patch?:boolean, delete?:boolean},
  authentication: AuthType | CrudAuthAll | CrudAuthSome;
  manageFields?:ManageableFields;
  dataType: ObjectType;
  search?: SearchTypes;
  description?: string;
  preferredImplementation?: {type: 'elasticsearch'; index: string};
};

export type OutputSuccess = {
  name: string;
  authentication: AuthType;
  manageFields: ManageableFields;
  method: HttpMethods;
  arguments: ObjectType;
  returns: ObjectType;
  preferredImplementation?: Elastic | KeyValue;
  search?: SearchTypes;
};

export type Output =
  | {type: 'result'; key: string; results: OutputSuccess[]; errors?: undefined;}
  | {type: 'error'; key?: undefined; errors: (Ajv.ErrorObject | string)[] | string; results?: undefined;};

export const baseSchemaLocation = `${dirname(fileURLToPath(import.meta.url))}/../../schema/`
