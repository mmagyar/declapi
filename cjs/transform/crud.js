"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = void 0;
const microtil_1 = require("microtil");
const jsonSchema_js_1 = require("./jsonSchema.js");
const util_js_1 = require("../util.js");
const contractOptions = (input) => {
    if (Array.isArray(input)) {
        if (input.some(x => x === '?')) {
            return input;
        }
        return input.concat(['?']);
    }
    return [input, '?'];
};
const searchToType = (idType, dataType, search) => {
    if (search === 'idOnly') {
        return { id: [idType, { $array: idType }, '?'] };
    }
    else if (search === 'textSearch') {
        return { search: ['string', '?'], id: [idType, { $array: idType }, '?'] };
    }
    else if (search === 'full') {
        return microtil_1.map(dataType, value => contractOptions(value));
    }
    else if (!search) {
        return {};
    }
    return search;
};
const checkIdField = (contract) => {
    if (contract.dataType.id === undefined) {
        return {
            type: 'error',
            errors: 'id field does not exist in the data declaration'
        };
    }
    const idType = contract.dataType.id;
    if (!(idType === 'string' || idType.$string)) {
        return {
            type: 'error',
            errors: 'Type of id field must be string'
        };
    }
    return false;
};
const checkManageFields = (contract) => {
    for (const [key, value] of Object.entries(contract.manageFields || {})) {
        if (value) {
            const fieldType = contract.dataType[key];
            if (fieldType === undefined) {
                return {
                    type: 'error',
                    errors: `managed field "${key}" is not present on data type`
                };
            }
            if (!(fieldType === 'string' || fieldType.$string)) {
                return {
                    type: 'error',
                    errors: `managed field "${key}" must be a string, current type :${fieldType}`
                };
            }
        }
    }
    return false;
};
const removeManaged = (args, manageFields) => {
    const result = Object.assign({}, args);
    for (const [key, value] of Object.entries(manageFields || {})) {
        if (value) {
            delete result[key];
        }
    }
    return result;
};
const isCrudAuth = (tbd) => tbd.post !== undefined;
const isCrudAuthSome = (tbd) => tbd.modify !== undefined;
const transformForPost = (tbd) => Array.isArray(tbd) && tbd.find(x => x.createdBy) ? true : tbd;
exports.transform = (data) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e;
    const valid = yield jsonSchema_js_1.validate(yield util_js_1.loadJSON(`${util_js_1.baseSchemaLocation}crudContractSchema.json`), data);
    if (jsonSchema_js_1.isValidationError(valid))
        return valid;
    const contractData = data;
    const au = contractData.authentication;
    const auth = {
        GET: isCrudAuth(au) ? au.get : (isCrudAuthSome(au) ? au.get : au),
        POST: isCrudAuth(au) ? au.post : transformForPost((isCrudAuthSome(au) ? au.modify : au)),
        PUT: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
        PATCH: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
        DELETE: isCrudAuth(au) ? au.delete : (isCrudAuthSome(au) ? au.delete || au.modify : au)
    };
    const createOutput = (method, args, returns = contractData.dataType) => ({
        method,
        name: contractData.name,
        authentication: auth[method],
        manageFields: contractData.manageFields || {},
        search: contractData.search,
        preferredImplementation: contractData.preferredImplementation,
        arguments: args,
        returns
    });
    const returnArray = { $array: contractData.dataType };
    const output = [];
    const errorWithId = checkIdField(contractData);
    if (errorWithId)
        return errorWithId;
    const errorWithManageFields = checkManageFields(contractData);
    if (errorWithManageFields)
        return errorWithManageFields;
    const idType = contractData.dataType.id;
    if (((_a = contractData.methods) === null || _a === void 0 ? void 0 : _a.get) !== false) {
        const search = contractData.search;
        output.push(createOutput('GET', searchToType(idType, contractData.dataType, search), returnArray));
    }
    if (((_b = contractData.methods) === null || _b === void 0 ? void 0 : _b.post) !== false) {
        const post = Object.assign(Object.assign({}, contractData.dataType), { id: [idType, '?'] });
        output.push(createOutput('POST', removeManaged(post, contractData.manageFields)));
    }
    if (((_c = contractData.methods) === null || _c === void 0 ? void 0 : _c.put) !== false) {
        output.push(createOutput('PUT', removeManaged(contractData.dataType, contractData.manageFields)));
    }
    if (((_d = contractData.methods) === null || _d === void 0 ? void 0 : _d.patch) !== false) {
        const patch = Object.assign(Object.assign({}, microtil_1.map(contractData.dataType, contractOptions)), { id: idType });
        output.push(createOutput('PATCH', removeManaged(patch, contractData.manageFields)));
    }
    if (((_e = contractData.methods) === null || _e === void 0 ? void 0 : _e.delete) !== false) {
        const deleteIds = { id: [idType, { $array: idType }] };
        output.push(createOutput('DELETE', deleteIds, returnArray));
    }
    return { type: 'result', key: contractData.name, results: output };
});
exports.default = exports.transform;
//# sourceMappingURL=crud.js.map