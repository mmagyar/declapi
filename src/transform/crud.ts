import { ValueType, ObjectType, StringType } from 'yaschva'
import { map } from 'microtil'
import { validate, isValidationError } from './jsonSchema'
import {
  CrudContract,

  CrudAuthAll,
  CrudAuthSome, OutputSuccess, Output, baseSchemaLocation, ManageableFields
} from './types'
import {
  HttpMethods,
  SearchTypes
} from 'declarapi-runtime'

const contractOptions = (input: ValueType | ValueType[]): ValueType[] => {
  if (Array.isArray(input)) {
    if (input.some(x => x === '?')) { return input }

    return input.concat(['?'])
  }
  return [input, '?']
}

const searchToType =
  (idType: 'string'| StringType, dataType: ObjectType, search?: SearchTypes): ObjectType => {
    if (search === 'idOnly') {
      return { id: [idType, { $array: idType }, '?'] }
    } else if (search === 'textSearch') {
      return { search: ['string', '?'], id: [idType, { $array: idType }, '?'] }
    } else if (search === 'full') {
      return map(dataType, value => contractOptions(value))
    } else if (!search) {
      return {}
    }

    return search
  }

const checkIdField = (contract:CrudContract) :Output | false => {
  if (contract.dataType.id === undefined) {
    return {
      type: 'error',
      errors: 'id field does not exist in the data declaration'
    }
  }

  const idType: any = contract.dataType.id
  if (!(idType === 'string' || idType.$string)) {
    return {
      type: 'error',
      errors: 'Type of id field must be string'
    }
  }

  return false
}
const checkManageFields = (contract: CrudContract): Output|false => {
  for (const [key, value] of Object.entries(contract.manageFields || {})) {
    if (value) {
      const fieldType: any = contract.dataType[key]
      if (fieldType === undefined) {
        return {
          type: 'error',
          errors: `managed field "${key}" is not present on data type`
        }
      }

      if (!(fieldType === 'string' || fieldType.$string)) {
        return {
          type: 'error',
          errors: `managed field "${key}" must be a string, current type :${fieldType}`
        }
      }
    }
  }

  return false
}

const removeManaged = (args: ObjectType, manageFields?: ManageableFields):ObjectType => {
  const result = { ...args }
  for (const [key, value] of Object.entries(manageFields || {})) {
    if (value) {
      delete result[key]
    }
  }

  return result
}
const isCrudAuth = (tbd: any): tbd is CrudAuthAll => tbd.post !== undefined
const isCrudAuthSome = (tbd: any): tbd is CrudAuthSome => tbd.modify !== undefined
const transformForPost = (tbd: any) => Array.isArray(tbd) && tbd.find(x => x.createdBy) ? true : tbd
export const transform = async (data:CrudContract | any): Promise<Output> => {
  const valid = await validate(require(`${baseSchemaLocation}crudContractSchema.json`), data)
  if (isValidationError(valid)) return valid

  const contractData: CrudContract = data

  const au = contractData.authentication

  const auth = {
    GET: isCrudAuth(au) ? au.get : (isCrudAuthSome(au) ? au.get : au),
    POST: isCrudAuth(au) ? au.post : transformForPost((isCrudAuthSome(au) ? au.modify : au)),
    PUT: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    PATCH: isCrudAuth(au) ? au.put : (isCrudAuthSome(au) ? au.modify : au),
    DELETE: isCrudAuth(au) ? au.delete : (isCrudAuthSome(au) ? au.delete || au.modify : au)
  }

  const createOutput = (method: HttpMethods, args: ObjectType,
    returns: ObjectType = contractData.dataType): OutputSuccess => ({
    method,
    name: contractData.name,
    authentication: auth[method],
    manageFields: contractData.manageFields || {},
    search: contractData.search,
    preferredImplementation: contractData.preferredImplementation,
    arguments: args,
    returns
  })

  const returnArray = { $array: contractData.dataType }
  const output: OutputSuccess[] = []

  const errorWithId = checkIdField(contractData)
  if (errorWithId) return errorWithId

  const errorWithManageFields = checkManageFields(contractData)
  if (errorWithManageFields) return errorWithManageFields

  const idType: any = contractData.dataType.id
  if (contractData.methods?.get !== false) {
    const search = contractData.search
    output.push(createOutput('GET',
      searchToType(idType, contractData.dataType, search), returnArray))
  }

  if (contractData.methods?.post !== false) {
    const post = { ...contractData.dataType, id: [idType, '?'] }
    output.push(createOutput('POST', removeManaged(post, contractData.manageFields)))
  }

  if (contractData.methods?.put !== false) {
    output.push(createOutput('PUT', removeManaged(contractData.dataType, contractData.manageFields)))
  }

  if (contractData.methods?.patch !== false) {
    const patch: {[s: string]: ValueType | ValueType[];} =
    { ...map(contractData.dataType, contractOptions), id: idType }
    output.push(createOutput('PATCH', removeManaged(patch, contractData.manageFields)))
  }

  if (contractData.methods?.delete !== false) {
    const deleteIds: {[s: string]: ValueType[];} = { id: [idType, { $array: idType }] }
    output.push(createOutput('DELETE', deleteIds, returnArray))
  }

  return { type: 'result', key: contractData.name, results: output }
}

export default transform
