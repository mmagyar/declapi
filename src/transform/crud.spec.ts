import { transform } from './crud'
import { CrudContract, CrudAuthAll, CrudAuthSome } from './types'
describe('transform crud', () => {
  it('id must be present on input', async () => {
    const resultErr = await transform({ name: 'test', authentication: false, dataType: { notId: 'string' } })
    expect(resultErr).toStrictEqual({
      type: 'error',
      errors: 'Field with the name set for idFieldName: id does not exist in the data declaration'
    })

    const result = await transform({
      name: 'test',
      authentication: false,
      dataType: { id: 'string', notId: 'string' }
    })
    expect(result).toHaveProperty('type', 'result')
    expect(result).toHaveProperty('results')
    expect(result.results).toHaveLength(5)
  })

  it('generates a full search option', async () => {
    const input:CrudContract = {
      name: 'test',
      authentication: false,
      dataType: {
        id: 'string',
        myNumber: 'number',
        myString: 'string'
      },
      search: 'full'
    }
    expect(await transform(input)).toMatchSnapshot()
  })

  it('generates an id only get', async () => {
    const input:CrudContract = {
      name: 'test',
      authentication: false,
      dataType: {
        id: 'string',
        myNumber: 'number',
        myString: 'string'
      },
      search: 'idOnly'
    }

    expect(await transform(input)).toMatchSnapshot()
  })

  it('generates a custom arguments parameters for get ', async () => {
    const input:CrudContract = {
      name: 'test',
      authentication: false,
      dataType: {
        id: 'string',
        myNumber: 'number',
        myString: 'string'
      },
      search: { customSearchField: 'string' }
    }
    const result = await transform(input)
    expect(result.results?.[0]?.arguments?.customSearchField).toEqual('string')
    expect(result).toMatchSnapshot()
  })

  it('returns an object with an error message on invalid id name', async () => {
    const input:CrudContract = {
      name: 'test',
      idFieldName: 'notAnId',
      authentication: false,
      dataType: {
        id: 'string',
        notAnId: 'boolean'
      },
      search: { customSearchField: 'string' }
    }
    const result = await transform(input)
    expect(result.errors).toEqual('Type of id field must be string')
  })

  it('returns an object with an error message on validation error', async () => {
    const input: any = {
      name: 'test',
      authenticationz: false,
      dataType: {
        id: 'string',
        notAnId: 'boolean'
      }
    }
    const result = await transform(input)
    expect(result.errors).toStrictEqual([{ dataPath: '', keyword: 'additionalProperties', message: 'should NOT have additional properties', params: { additionalProperty: 'authenticationz' }, schemaPath: '#/additionalProperties' }])
  })

  it('makes all parameters optional for patch', async () => {
    const input: CrudContract = {
      name: 'test',
      authentication: false,
      dataType: {
        id: 'string',
        notAnId: ['boolean', '?'],
        singleElementArrayType: ['string'],
        obj: { a: 'string', b: 'number' },
        obj2: { a: ['string', '?'], b: ['number', '?'] },
        duoType: ['boolean', 'number']
      }
    }
    const result = await transform(input)
    expect(result.results?.find(x => x.method === 'patch')?.arguments).toStrictEqual({
      id: 'string',
      notAnId: ['boolean', '?'],
      singleElementArrayType: ['string', '?'],
      obj: [{ a: 'string', b: 'number' }, '?'],
      obj2: [{ a: ['string', '?'], b: ['number', '?'] }, '?'],
      duoType: ['boolean', 'number', '?']
    })
  })

  it('supports different auth protocols', async () => {
    const withAuth: (auth? : string[] | boolean | CrudAuthAll | CrudAuthSome) => CrudContract =
     (auth: string[] | boolean | CrudAuthAll | CrudAuthSome = true) => ({
       name: 'test',
       authentication: auth,
       dataType: {
         id: 'string',
         notAnId: 'boolean'
       }
     })

    expect((await transform(withAuth(false))).results?.every(x => x.authentication === false)).toBeTruthy()
    expect((await transform(withAuth(true))).results?.every(x => x.authentication === true)).toBeTruthy()

    const stringAllRole = (await transform(withAuth(['aUserRole']))).results || []
    expect(stringAllRole).toHaveLength(5)
    stringAllRole.forEach(x => expect(x.authentication).toStrictEqual(['aUserRole']))

    const resultAuthSome = await transform(withAuth({ get: false, modify: ['admin'] }))
    expect(resultAuthSome.results?.find(x => x.method === 'get')?.authentication).toStrictEqual(false)
    const authSomeOtherMethods = resultAuthSome.results?.filter(x => x.method !== 'get') || []
    expect(authSomeOtherMethods).toHaveLength(4)
    authSomeOtherMethods.forEach(x => expect(x.authentication).toStrictEqual(['admin']))

    const resultAuth = await transform(withAuth({ get: false, post: true, put: ['owner'], delete: ['admin'] }))
    expect(resultAuth.results?.find(x => x.method === 'get')?.authentication).toStrictEqual(false)
    expect(resultAuth.results?.find(x => x.method === 'post')?.authentication).toStrictEqual(true)
    expect(resultAuth.results?.find(x => x.method === 'put')?.authentication).toStrictEqual(['owner'])
    expect(resultAuth.results?.find(x => x.method === 'delete')?.authentication).toStrictEqual(['admin'])
  })
})
