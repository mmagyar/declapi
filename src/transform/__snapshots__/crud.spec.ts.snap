// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform crud does not generate arguments for get is search is not set 1`] = `
Object {
  "key": "test",
  "results": Array [
    Object {
      "arguments": Object {},
      "authentication": false,
      "manageFields": Object {},
      "method": "GET",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": undefined,
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "POST",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": undefined,
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PUT",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": undefined,
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PATCH",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": undefined,
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "DELETE",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": undefined,
    },
  ],
  "type": "result",
}
`;

exports[`transform crud generates a custom arguments parameters for get  1`] = `
Object {
  "key": "test",
  "results": Array [
    Object {
      "arguments": Object {
        "customSearchField": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "GET",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": Object {
        "customSearchField": "string",
      },
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "POST",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": Object {
        "customSearchField": "string",
      },
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PUT",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": Object {
        "customSearchField": "string",
      },
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PATCH",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": Object {
        "customSearchField": "string",
      },
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "DELETE",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": Object {
        "customSearchField": "string",
      },
    },
  ],
  "type": "result",
}
`;

exports[`transform crud generates a full text search search option 1`] = `
Object {
  "key": "test",
  "results": Array [
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
          "?",
        ],
        "search": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "GET",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "textSearch",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "POST",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "textSearch",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PUT",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "textSearch",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PATCH",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "textSearch",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "DELETE",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "textSearch",
    },
  ],
  "type": "result",
}
`;

exports[`transform crud generates an id only get 1`] = `
Object {
  "key": "test",
  "results": Array [
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "GET",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "idOnly",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "POST",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "idOnly",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PUT",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "idOnly",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PATCH",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "idOnly",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "DELETE",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "idOnly",
    },
  ],
  "type": "result",
}
`;

exports[`transform crud generates full, parametric search option 1`] = `
Object {
  "key": "test",
  "results": Array [
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "GET",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "full",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          "?",
        ],
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "POST",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "full",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PUT",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "full",
    },
    Object {
      "arguments": Object {
        "id": "string",
        "myNumber": Array [
          "number",
          "?",
        ],
        "myString": Array [
          "string",
          "?",
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "PATCH",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "id": "string",
        "myNumber": "number",
        "myString": "string",
      },
      "search": "full",
    },
    Object {
      "arguments": Object {
        "id": Array [
          "string",
          Object {
            "$array": "string",
          },
        ],
      },
      "authentication": false,
      "manageFields": Object {},
      "method": "DELETE",
      "name": "test",
      "preferredImplementation": undefined,
      "returns": Object {
        "$array": Object {
          "id": "string",
          "myNumber": "number",
          "myString": "string",
        },
      },
      "search": "full",
    },
  ],
  "type": "result",
}
`;
